src/io/PerCellGenerator.cpp

```diff
@@ -0,0 +1,166 @@
+#include "PerCellGenerator.h"
+
+#include <iostream>
+#include <random>
+
+#include "Domain.h"
+#include "IOHelpers.h"
+#include "Simulation.h"
+#include "ensemble/EnsembleBase.h"
+#include "molecules/Molecule.h"
+#include "parallel/DomainDecompBase.h"
+#include "utils/Logger.h"
+#include "utils/xmlfileUnits.h"
+
+unsigned long PerCellGenerator::readPhaseSpace(ParticleContainer *particleContainer, Domain *domain,
+											   DomainDecompBase *domainDecomp) {
+	auto &componentVector = *global_simulation->getEnsemble()->getComponents();
+	mardyn_assert(not componentVector.empty());
+	Component *component = &(componentVector[0]);
+	// The halo should not be filled when using readPhaseSpace, as the haloExchange is handled by the domain
+	// decomposition!
+	bool fillHalo = false;
+	fillContainer(particleContainer, component, _numMoleculesPerCell, fillHalo);
+	if (_numMoleculesPerCell == 0 and _generateAtLeastTwoParticles) {
+		// There should always be at least two particles!
+		generateTwoParticles(particleContainer, component);
+	}
+	unsigned long totalNumParticles =
+		IOHelpers::makeParticleIdsUniqueAndGetTotalNumParticles(particleContainer, domainDecomp);
+	IOHelpers::initializeVelocityAccordingToTemperature(particleContainer, domainDecomp, _initTemperature);
+	IOHelpers::removeMomentum(particleContainer, componentVector, domainDecomp);
+	domain->setglobalNumMolecules(totalNumParticles);
+	return totalNumParticles;
+}
+
+void PerCellGenerator::readXML(XMLfileUnits &xmlconfig) {
+	Log::global_log->info() << "------------------------------------------------------------------------" << std::endl;
```

Comment:
while I like visual breakups this does not seem to be in the output style of ls1?

Reply:
there are a couple of places where this is used, so I'll leave it.